scripts/apps/archive/controllers/UploadController.ts

```diff
@@ -258,20 +259,30 @@ export function UploadController(
     };
 
     $scope.upload = function() {
-        var promises = [];
+        if (isEmpty($scope.items)) {
+            return Promise.resolve();
+        }
 
-        _.each($scope.items, (item) => {
+        // upload items in sequence, and resolve when all are done
+        return serializePromises($scope.items.map((item) => {
             if (!item.model && !item.progress) {
                 item.upload = null;
-                promises.push(uploadFile(item));
+                return () => uploadFile(item);
             }
-        });
-        if (promises.length) {
-            return $q.all(promises);
-        }
-        return $q.when();
+
+            return () => Promise.resolve(item);
+        }));
     };
 
+    function serializePromises(callbacks) {
```

Comment:
rename `callbacks` to `promiseCreators` or something else more exact. I'd also add typings:

```
function serializePromises<T>(promiseCreators: Array<() => Promise<T>>): Promise<Array<T>> {
    let promise = Promise.resolve(null);

    return Promise.all(promiseCreators.map((callback) => {
        promise = promise.then(callback);
        return promise;
    }));
}
```